src/SmoFacade/Listener.cs

```diff
@@ -135,14 +135,52 @@ private static string AgListenerName(string dataSource)
     {
       var parts = agInstanceName.Split('\\');
       if(parts.Length == 1)
-        connBuilder.DataSource = Dns.GetHostEntry(agInstanceName).HostName;
+        connBuilder.DataSource = ResolveDnsAndGetHostName(connBuilder.DataSource, agInstanceName);
       else if(parts.Length == 2)
         // NamedInstances: chop instance name, resolve DNS, slap instance name back on!
-        connBuilder.DataSource = $"{Dns.GetHostEntry(parts[0]).HostName}\\{parts[1]}";
+        connBuilder.DataSource = $"{ResolveDnsAndGetHostName(connBuilder.DataSource, parts[0])}\\{parts[1]}";
       else
         throw new ArgumentException($"agInstanceName param {agInstanceName} cannot be resolved by DNS");
 
       return new Server(connBuilder.ToString(), credentialName);
     }
+
+    /// <summary>
+    ///   Resolves "agReplicaInstanceName" to a FQDNS name
+    ///   However on Unix OS, "agReplicaInstanceName" should be a complete domain - "{sub domain}.{second level domain}.{...}.{top level domain}" (eg: "xyz.abc.def.com")
+    ///   Therefore, we use {second-level} -> {top-level} part of the domain from "agListenerDomain" and append it to "agReplicaInstanceName"
+    /// </summary>
+    /// <param name="agListenerDomain">The complete domain for the SQL server AG listener</param>
+    /// <param name="agReplicaInstanceName">The specific name of a replica instance within the AG</param>
+    private static string ResolveDnsAndGetHostName(string agListenerDomain, string agReplicaInstanceName)
+    {
+      Tuple<string, string> domain_port = SplitDomainPort(agListenerDomain);
+      if (Environment.OSVersion.Platform == PlatformID.Unix)
+      {
+        string[] domainFragments = domain_port.Item1.Split('.');
+        domainFragments[0] = "";

```

Comment:
Why are you setting domainFragments[0] to ""?  This seems random... would like a comment here or soemthing

Reply:
sure will put in a comment there

It's to strip off the first part of the domain before appending it to the instance name
so if it's like `abc.def.ghi` we want only `.def.ghi` to be appended to the instance name

I think we should end up using `UriBuilder` class in the end rather than our own string manipulations:
https://docs.microsoft.com/en-us/dotnet/api/system.uribuilder?view=netcore-2.0